### DISCUSSION TITLE
Morse code using OOK

### DISCUSSION THREAD

**Discussion by Sumya99** [jgromes/RadioLib#448] [Q&A]
Hello sir, 

First, I am a quite novice in this RF field and hope you help me. I am trying to transmit Morse code using OOK modulation scheme on RFM95PW. Is it possible using this library? 

Thanks in advance

**Comment by jgromes** [comment]
Morse code is by definition on-off keying - it's just that it encodes data into the length of the "on" beeps, rather than the presence of them, as standard OOK does.

In short: yes, you can, have you tried it? There's examples.

**Reply by Sumya99** [reply]
Thank you for the reply. Yes, I have used that examples in Morse code file. There is a definition which says that example is working in FSK or AFSK mode. And tried to set OOK modulation true, but -107 error occured "the current modulation is invalid for requested operation". Can that examples only work in FSK-CW ?

**Reply by jgromes** [reply]
In theory, OOK should work, could you post the code that's failing? There's also an example that transmit just simple audio tones in OOK mode (https://github.com/jgromes/RadioLib/blob/master/examples/AFSK/AFSK_Tone_AM/AFSK_Tone_AM.ino).

The difference between FSK Morse and AFSK Morse is just in the way audio tones are transmitted. In FSK mode, you have to demodulate in SSB mode to get the audio. In AFSK mode, you have to demodulate in NFM mode. If you set the module to OOK, then you will have to use AM to demodulate.

**Reply by Sumya99** [reply]
Here is the code I tried FSK mode to OOK. Could you please review ?

```c++
// include the library
#include <RadioLib.h>

// SX1278 has the following connections:
// NSS pin:   10
// DIO0 pin:  2
// RESET pin: 9
// DIO1 pin:  3
SX1278 radio = new Module(10, 2, 9, 3);

// or using RadioShield
// https://github.com/jgromes/RadioShield
//SX1278 radio = RadioShield.ModuleA;

// create AFSK client instance using the FSK module
// pin 5 is connected to SX1278 DIO2
AFSKClient audio(&radio, 5);

// create Morse client instance using the AFSK instance
MorseClient morse(&audio);

void setup() {
  Serial.begin(9600);
  // initialize SX1278 with default settings
  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK();
  state = radio.setFrequency(437.135);
 
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  **// It is where I tried to make a change FSK to OOK
  state = radio.setOOK(true);
  state = radio.setDataShapingOOK(1);
  if (state != RADIOLIB_ERR_NONE) {
    Serial.print(F("Unable to change modulation, code "));
    Serial.println(state);
    while (true);
  }**
  
  // initialize Morse client
  Serial.print(F("[Morse] Initializing ... "));
  // AFSK tone frequency:         400 MHz
  // speed:                       20 words per minute
  state = morse.begin(400);
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }
}

void loop() {
  Serial.print(F("[Morse] Sending Morse data ... "));
  morse.startSignal();
  morse.print("A");
  delay(1000);
  Serial.println(F("done!"));
}
```
